### PR TITLE
Prebid Core: refactor bidderSettings to have only one entry point

### PR DESCRIPTION
**Pull request by dgirardi**
## Type of change

- [x] Refactoring (no functional changes, no api changes)

## Description of change

Add thin wrapper around `$$PREBID_GLOBAL$$.bidderSettings` that reduces some duplicate logic around fallback to `bidderSettings.standard`

This was inspired by https://github.com/prebid/Prebid.js/issues/7280, which proposes using `bidderSettings` for a new purpose.

Note that I found some horrors lurking deep in the darkness; see the PR comments on the code changes.

### CODE REVIEW

File: test/spec/modules/sortableAnalyticsAdapter_spec.js [hunk 1]
```diff
@@ -203,11 +202,11 @@ describe('Sortable Analytics Adapter', function() {
         brc: 1,
         brid: ['10141593b1d84a', '37a8760be6db23'],
         rs: ['300x250', '728x90'],
-        btcp: [0.70, 0.50],
+        btcp: [0.70 * 0.95, 0.50 * 0.95],
         btcc: 'USD',
         btin: true,
         btsrc: 'sortable',
-        c: [0.70, 0.50],
+        c: [0.70 * 0.95, 0.50 * 0.95],
```
Comment (dgirardi) [inline test/spec/modules/sortableAnalyticsAdapter_spec.js:209]: this suite sets up `bidCpmAdjustment` by mocking `getGlobal`, which was not seen by `adjustBids` in `auction.js` because that looked directly at `$$PREBID_GLOBAL$$.bidderSettings`. This test never caught it, but there are other assertions in this suite that do expect numbers to get through the adjustment, so we need to modify either the setup or the assertion.

File: src/auction.js [hunk 2]
```diff
@@ -661,14 +662,13 @@ export function getStandardBidderSettings(mediaType, bidderCode) {
...
 
-  let bidderSettings = $$PREBID_GLOBAL$$.bidderSettings;
-  if (!bidderSettings[CONSTANTS.JSON_MAPPING.BD_SETTING_STANDARD]) {
-    bidderSettings[CONSTANTS.JSON_MAPPING.BD_SETTING_STANDARD] = {};
-  }
-  if (!bidderSettings[CONSTANTS.JSON_MAPPING.BD_SETTING_STANDARD][CONSTANTS.JSON_MAPPING.ADSERVER_TARGETING]) {
-    bidderSettings[CONSTANTS.JSON_MAPPING.BD_SETTING_STANDARD][CONSTANTS.JSON_MAPPING.ADSERVER_TARGETING] = [
+  if (!standardSettings[CONSTANTS.JSON_MAPPING.ADSERVER_TARGETING]) {
+    standardSettings[CONSTANTS.JSON_MAPPING.ADSERVER_TARGETING] = [
+      // NOTE: some callers (probably tests) are modifying these - they are not safe to cache
```
Comment (dgirardi) [inline src/auction.js:671]: I tried generating these once as the original version did, but someone is messing with the return value. I was not able to find who that is or why that was not a problem before my change.

File: src/auction.js [hunk 3]
```diff
@@ -691,7 +692,7 @@ export function getStandardBidderSettings(mediaType, bidderCode) {
     });
 
     // Adding hb_cache_host
-    if (config.getConfig('cache.url') && (!bidderCode || deepAccess(bidderSettings, `${bidderCode}.sendStandardTargeting`) !== false)) {
+    if (config.getConfig('cache.url') && (!bidderCode || bidderSettings.get(bidderCode, 'sendStandardTargeting') !== false)) {
```
Comment (dgirardi) [inline src/auction.js:695]: This is potentially a breaking change: previously, this would be evaluated the first time this function was called with `mediaType === 'video'` and would set `hb_cache_host` globally after that. There was also no fallback to `bidderSettings.standard`.

File: src/auction.js [hunk 4]
```diff
@@ -711,19 +712,15 @@ export function getKeyValueTargetingPairs(bidderCode, custBidObj, bidReq) {
...
-      setKeys(keyValues, bidderSettings[bidderCode], custBidObj, bidReq);
-      custBidObj.sendStandardTargeting = bidderSettings[bidderCode].sendStandardTargeting;
-    }
+  const standardSettings = getStandardBidderSettings(custBidObj.mediaType, bidderCode);
+  setKeys(keyValues, standardSettings, custBidObj, bidReq);
+
+  // 2) set keys from specific bidder setting override if they exist
+  if (bidderCode && bidderSettings.getOwn(bidderCode, CONSTANTS.JSON_MAPPING.ADSERVER_TARGETING)) {
+    setKeys(keyValues, bidderSettings.ownSettingsFor(bidderCode), custBidObj, bidReq);
+    custBidObj.sendStandardTargeting = bidderSettings.get(bidderCode, 'sendStandardTargeting');
```
Comment (dgirardi) [inline src/auction.js:723]: Another potentially breaking change: the logic here was a bit confusing, copying `sendStandardTargeting` only if `adserverTargeting` is defined, which was also overriding `standard.sendStandardTargeting`. From the public documentation it seems like we should be using the `standard` setting if there is bidder-specific one.

File: src/auction.js (hunk 3 above)
Comment (robertrmartinez) [reply src/auction.js:700]: Breaking change... more like a bug fix!

File: src/auction.js (hunk 2 above)
Comment (robertrmartinez) [reply src/auction.js]: This was something I recently fixed.

https://github.com/prebid/Prebid.js/pull/7607

The default way this function is called is through https://github.com/prebid/Prebid.js/blob/4c424a06883a421bb25dd96d1cf8fbda90741e9c/src/auction.js#L674

This always passes undefined. And for these KVP sets it needs access to the bidReq.mediaType.

So when KVP's are set, it actually always runs the OR `getPriceGranularity(bid.mediaType, bidReq);`

Which will in turn select the right granularity.

There are only two other places which use this function, and they directly pass in the granularity, which is why we had to support the `granularity || granularity`

https://github.com/prebid/Prebid.js/blob/master/modules/adpod.js#L126

https://github.com/prebid/Prebid.js/blob/master/modules/konduitWrapper.js#L104

So this is actually fine with the current way this is setup, but yours is probably smarter.

Also the entire Targeting logic is crazy and needs to be rewritten! haha.

**Comment by robertrmartinez** [comment]
Thanks @dgirardi 

A Much needed update.

I will take some time to run some tests which use Bidder Settings and do a formal review sometime this week.

It is sort of a big change, so we should make sure to cover all the bases!

Lots of publishers rely on utilizing `pbjs.bidderSettings` and modifying it themselves.

File: src/auction.js (hunk 2 above)
Comment (dgirardi) [reply src/auction.js]: I don't think it's fine in general: the `|| getPriceGranularity(bid.mediaType, bidReq)` is only evaluated on the first call because `granularity` is in the closure scope. See the test case I added in `auctionmanager_spec`, that fails without this change.

Comment (robertrmartinez) [reply src/auction.js]: I get it. Thanks!

**Review by Rothalack — APPROVED** [review]
This looks good to me, working in testing.

I think the documentation is still accurate other than maybe we should add mention of caching to that documentation.

File: src/utils.js [hunk 5]
```diff
@@ -1299,5 +1299,49 @@ export function cyrb53Hash(str, seed = 0) {
...
+  return function () {
+    return f(null);
+  }
+}
+
+/**
+ * Return an object that is the result of recursively merging properties from the given objects. If multiple
+ * objects define the same "leaf" properties, the right-most's value is picked.
+ */
+export function deepMerge(left, right) {
```
Comment (dgirardi) [inline src/utils.js:1332]: I just learned that there is already a `mergeDeep` in here. I'll post a revision early next week to slim this down some.

**Comment by ChrisHuie** [comment]
@robertrmartinez can you look at this pr when you get a chance 🙏

**Comment by Rothalack** [comment]
Hey @robertrmartinez , will you be able to review this PR? Don't want to let it closed.

File: src/utils.js [hunk 6]
```diff
@@ -1326,5 +1326,28 @@ export function cyrb53Hash(str, seed = 0) {
...
   const bidderSettings = getGlobal().bidderSettings;
   return ((bidderSettings[bidderCode] && bidderSettings[bidderCode].allowZeroCpmBids === true) ||
-   (bidderSettings.standard && bidderSettings.standard.allowZeroCpmBids === true));
+    (bidderSettings.standard && bidderSettings.standard.allowZeroCpmBids === true));
+}
+
+/**
+ * Return a memoized version of the given function that uses the first argument as cache key.
+ */
+export function memoize1(fun) {
```
Comment (FilipStamenkovic) [inline src/utils.js:1335]: Why is this called `memoize1`?  Why not just `memoize`?

File: src/utils.js [hunk 7]
```diff
@@ -1326,5 +1326,28 @@ export function cyrb53Hash(str, seed = 0) {
...
+      cache[a0] = fun(a0, ...rest);
+    }
+    return cache[a0];
+  }
+}
+
+/**
+ * Return a memoized version of the given zero-arg function.
+ */
+export function cache(fun) {
```
Comment (FilipStamenkovic) [inline src/utils.js:1348]: Do we need both cache and memoize?  
It seems to me that just `memoize` function is enough?

**Review by FilipStamenkovic — CHANGES_REQUESTED** [review]
This pull request has conflicts. Can you pull/rebase your branch and resolve this conflicts? Also, take a look at my other comments.

File: src/bidderSettings.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
+import {deepAccess, mergeDeep} from './utils.js';
+import {getGlobal} from './prebidGlobal.js';
+
+const CONSTANTS = require('./constants.json');
+
+export function ScopedSettings(getSettings, defaultScope) {
```
[the remark this answers is not in the document]
Comment (osazos) [reply src/bidderSettings.js:6]: Is there a reason to not using Class if you choose OOP?

[the remark this answers is not in the document]
Comment (dgirardi) [reply src/bidderSettings.js:6]: > Is there a reason to not using Class if you choose OOP?

Because when I started working on this IE11 was still a target :) I don't have a strong preference on the syntax, I can rewrite using `class` - should I take your comment as an encouragement?

> I'm not big fan of OOP concept.

Then some day I'll have to explain to you why you're wrong.

But more seriously, I don't understand how you can say we barely use OOP - config, auctionManager, auction, adapterManager, adapters, etc are all objects with capabilities. Is the problem that I use `prototype` instead of hidden closures as "private" methods? or the fact that you have to use `new` for this function but not for the others?

File: src/utils.js (hunk 6 above)
Comment (dgirardi) [reply src/utils.js:1335]: because it only uses the first argument as the memoization key.

File: src/bidderSettings.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (osazos) [reply src/bidderSettings.js:6]: >  should I take your comment as an encouragement?

Yes, totally :)

File: src/auction.js [hunk 9]
```diff
@@ -642,47 +643,51 @@ export const getAdvertiserDomain = () => {
...
+      ? function (bidResponse, bidReq) {
+        return value(bidResponse, bidReq);
+      }
+      : function (bidResponse) {
+        return getValue(bidResponse, value);
+      }
+  };
+}
+
+const defaultAdserverTargeting = cache(() => {
```
[the remark this answers is not in the document]
Comment (dgirardi) [reply src/auction.js:660]: Both `getPriceByGranularity` and `getAdvertiserDomain` should return equivalent closures no matter what the config is (the return value is what looks it up) so caching should not be a problem.

But I do think there's a good argument to be made for not caching - I am doing it here because it was in the original code, which over time got broken because more stuff was added in a non-cacheable way (in `getStandardBidderSettings`). 

So now we still have to recalculate the "standard" settings for each auction, plus we have to be mindful of what is being cached to avoid making unintentional changes to global state. I doubt the added complexity is worth the tiny performance improvement, so I also vote for removing the cache.

File: .eslintrc.js [hunk 10]
```diff
@@ -23,10 +23,13 @@ module.exports = {
     'BROWSERSTACK_USERNAME': false,
     'BROWSERSTACK_KEY': false
   },
+  // use babel as parser for fancy syntax
+  parser: '@babel/eslint-parser',
```
Comment (FilipStamenkovic) [inline .eslintrc.js:27]: Why do we need `@babel/eslint-parser`?  
Looking at the npm description I see:

> ESLint's default parser and core rules only support the latest final ECMAScript standard and do not support experimental (such as new features) and non-standard (such as Flow or TypeScript types) syntax provided by Babel

Do we need these 'experimental and non-standard' syntax? Or am I missing something?

File: modules/appnexusBidAdapter.js [hunk 11]
```diff
@@ -292,7 +293,7 @@ export const spec = {
       serverResponse.tags.forEach(serverBid => {
         const rtbBid = getRtbBid(serverBid);
         if (rtbBid) {
-          const cpmCheck = (isAllowZeroCpmBidsEnabled(bidderRequest.bidderCode)) ? rtbBid.cpm >= 0 : rtbBid.cpm > 0;
+          const cpmCheck = (bidderSettings.get(bidderRequest.bidderCode, 'allowZeroCpmBids') === true) ? rtbBid.cpm >= 0 : rtbBid.cpm > 0;
```
Comment (FilipStamenkovic) [inline modules/appnexusBidAdapter.js:296]: From readability perspective, to me, personally, previous version was much clearer.  
You just execute some function to get the info you need (`isAllowZeroCpmBidsEnabled`).  
With this new approach code is longer and it's more error prone since you have to remember name of the property `allowZeroCpmBids`.  

But, I guess since that is used in 2,3 places in the entire prebid.js than it's fine.

File: package.json [hunk 12]
```diff
@@ -30,6 +30,7 @@
   },
   "devDependencies": {
     "@babel/core": "^7.8.4",
+    "@babel/eslint-parser": "^7.16.5",
```
Comment (FilipStamenkovic) [inline package.json:33]: I think adding new npm package should have been done in separate PR. The reason is not to pollute this PR with those changes.

**Review by FilipStamenkovic — APPROVED** [review]

File: .eslintrc.js (hunk 10 above)
Comment (dgirardi) [reply .eslintrc.js:27]: It's part of the `class` syntax, specifically field declaration and private (#) methods and fields. Btw I care about objects but not really about syntax, I don't mind going either way with this.

File: modules/appnexusBidAdapter.js (hunk 11 above)
Comment (dgirardi) [reply modules/appnexusBidAdapter.js:296]: I don't disagree but there's a problem with circular imports when we put everything in `utils`. (new `bidderSettings` needs `utils.deepClone`, and a few others)
